src/RawPOSIX/src/interface/mem.rs

```diff
@@ -0,0 +1,228 @@
+        let mut vmmap = cage.vmmap.write();
+        let result;
+        
+        // pick an address of appropriate size, anywhere
+        if addr as usize == 0 {
+            result = vmmap.find_map_space(rounded_length as u32 >> PAGESHIFT, 1);
+        } else {
+            // use address user provided as hint to find address
+            result = vmmap.find_map_space_with_hint(rounded_length as u32 >> PAGESHIFT, 1, addr as u32);
+        }
+
+        // did not find desired memory region
+        if result.is_none() {
+            return syscall_error(Errno::ENOMEM, "mmap", "no memory");
+        }
+
+        let space = result.unwrap();
+        useraddr = (space.start() << PAGESHIFT) as i32;
+    }
+
+    // TODO: validate useraddr (like checking whether within the program break)
+
+    flags |= MAP_FIXED as i32;
+
+    // either MAP_PRIVATE or MAP_SHARED should be set, but not both
+    if (flags & MAP_PRIVATE as i32 == 0) == (flags & MAP_SHARED as i32 == 0) {
+        return syscall_error(Errno::EINVAL, "mmap", "invalid flags");
+    }
+
+    let vmmap = cage.vmmap.read();
+
+    let sysaddr = vmmap.user_to_sys(useraddr);
+
+    drop(vmmap);
+
+    if rounded_length > 0 {
+        if flags & MAP_ANONYMOUS as i32 > 0 {
+            fildes = -1;
+        }
+
+        let result = cage.mmap_syscall(sysaddr as *mut u8, rounded_length as usize, prot, flags, fildes, off);
+        
+        let vmmap = cage.vmmap.read();
+        let result = vmmap.sys_to_user(result);
+        drop(vmmap);
+
+        if result >= 0 {
+            if result != useraddr {
+                panic!("MAP_FIXED not fixed");
+            }
+
+            let mut vmmap = cage.vmmap.write();
+            let backing = {
+                MemoryBackingType::Anonymous
+                // TODO: should return backing type accordingly
+                // if flags & MAP_ANONYMOUS > 0 {
+                //     MemoryBackingType::Anonymous
+                // } else if flags & MAP_SHARED > 0 {
+
+                // }
+            };
+            vmmap.add_entry_with_overwrite((useraddr >> PAGESHIFT) as u32, (rounded_length >> PAGESHIFT) as u32, prot, 0, flags, backing, off, 0, cageid);
+        }
+    }
+
+    useraddr
+}
+
+pub fn sbrk_handler(cageid: u64, brk: u32) -> i32 {
+    let cage = cagetable_getref(cageid);
+
+    let mut vmmap = cage.vmmap.write();
+    let heap = vmmap.find_page(HEAP_ENTRY_INDEX).unwrap().clone();
+
+    if brk == 0 {
+        return (PAGESIZE * heap.npages) as i32;
+    }
+
+    let brk_page = ((brk + 65536 - 1) / 65536) * 16;
+
+    let heap_size = heap.npages;
+    vmmap.add_entry_with_overwrite(0, heap_size + brk_page, heap.prot, heap.maxprot, heap.flags, heap.backing, heap.file_offset, heap.file_size, heap.cage_id);
+    
+    let usr_heap_base = (heap_size * PAGESIZE) as i32;
+    let sys_heap_base = vmmap.user_to_sys(usr_heap_base)as *mut u8;
+
+    drop(vmmap);
+
+    // TODO: Currently we are not calling mmap to change prot here
```

Comment:
@qianxichen233 issue we need to fix. Need to grow memory to max size and protect as PROT_NONE, and then call mprotect here to give new brk memory correct permissions.